internal/engine/event_watch_manager_test.go

```diff
@@ -0,0 +1,131 @@
+	defer f.TearDown()
+
+	f.addManifest("someK8sManifest")
+
+	evt := &v1.Event{
+		Reason:  "because test",
+		Message: "hello world",
+	}
+
+	f.ewm.OnChange(f.ctx, f.store)
+	f.kClient.EmitEvent(f.ctx, evt)
+	expectedAction := store.K8SEventAction{Event: evt}
+	f.assertObservedK8sEventActions(expectedAction)
+
+}
+func TestEventWatchManager_needsWatchNoK8s(t *testing.T) {
+	f := newEWMFixture(t)
+	defer f.TearDown()
+
+	evt := &v1.Event{
+		Reason:  "because test",
+		Message: "hello world",
+	}
+
+	// No k8s manifests on the state, so OnChange shouldn't do anything --
+	// when we emit an event, we do NOT expect to see an action dispatched,
+	// since no watch should have been started.
+	f.ewm.OnChange(f.ctx, f.store)
+	f.kClient.EmitEvent(f.ctx, evt)
+	f.assertNoK8sEventActions()
+}
+
+type ewmFixture struct {
+	t          *testing.T
+	kClient    *k8s.FakeK8sClient
+	ewm        *EventWatchManager
+	ctx        context.Context
+	cancel     func()
+	store      *store.Store
+	getActions func() []store.Action
+
+	// old value of k8sEventsFeatureFlag env var, for teardown
+	// TODO(maia): remove this when we remove the feature flag
+	oldFeatureFlagVal string
+}
+
+func newEWMFixture(t *testing.T) *ewmFixture {
+	kClient := k8s.NewFakeK8sClient()
+
+	ctx := output.CtxForTest()
+	ctx, cancel := context.WithCancel(ctx)
+
+	ret := &ewmFixture{
+		kClient:           kClient,
+		ewm:               NewEventWatchManager(kClient),
+		ctx:               ctx,
+		cancel:            cancel,
+		t:                 t,
+		oldFeatureFlagVal: os.Getenv(k8sEventsFeatureFlag),
+	}
+
+	os.Setenv(k8sEventsFeatureFlag, "true")
+
+	ret.store, ret.getActions = store.NewStoreForTesting()
+	go ret.store.Loop(ctx)
+
+	return ret
+}
+
+func (f *ewmFixture) TearDown() {
+	_ = os.Setenv(k8sEventsFeatureFlag, f.oldFeatureFlagVal)
+	f.cancel()
+}
+
+func (f *ewmFixture) addManifest(manifestName string) {
+	state := f.store.LockMutableStateForTesting()
+	state.WatchFiles = true
+	dt := model.K8sTarget{Name: model.TargetName(manifestName)}
+	m := model.Manifest{Name: model.ManifestName(manifestName)}.WithDeployTarget(dt)
+	state.UpsertManifestTarget(store.NewManifestTarget(m))
+	f.store.UnlockMutableState()
+}
+
+func (f *ewmFixture) assertNoK8sEventActions() {
+	f.assertObservedK8sEventActions()
+}
+
+func (f *ewmFixture) assertObservedK8sEventActions(expectedActions ...store.K8SEventAction) {
+	if len(expectedActions) == 0 {
+		// assert no k8s event actions -- sleep briefly
+		// to give any actions a chance to get into the queue
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	start := time.Now()
+	for time.Since(start) < 200*time.Millisecond {
+		actions := f.getActions()
+		if len(actions) == len(expectedActions) {
+			break
+		}
+	}
+
+	var observedActions []store.K8SEventAction
+	for _, a := range f.getActions() {
+		sca, ok := a.(store.K8SEventAction)
+		if !ok {
+			f.t.Fatalf("got non-%T: %v", store.K8SEventAction{}, a)
+		}
+		observedActions = append(observedActions, sca)
```

Comment:
out of curiosity, why do you copy the actions to a new slice? i would have expected assert.Equal to work fine with f.getActions(), but maybe i'm misunderstanding

Reply:
ha this is copypasta from matt's code, so I don't know -- but i'm actually rejiggering it a bit because i changed it to `assert.Equal` but turned out that that didn't work with ErrorActions b/c of pointer nonsense.